pkg/rhai/progression/progression.go

```diff
@@ -473,64 +467,175 @@ func IsFinalStatusCaptured(trainJob *trainer.TrainJob) bool {
+					terminationStatus.EstimatedRemainingTimeSummary = "complete"
 				}
 			} else {
-				// For failed jobs: show progress context in summary
 				progressPct := 0
-				if annotationStatus.ProgressPercentage != nil {
-					progressPct = *annotationStatus.ProgressPercentage
+				if terminationStatus.ProgressPercentage != nil {
+					progressPct = *terminationStatus.ProgressPercentage
 				}
-				annotationStatus.EstimatedRemainingTimeSummary = fmt.Sprintf("failed at %d%%", progressPct)
+				terminationStatus.EstimatedRemainingTimeSummary = fmt.Sprintf("failed at %d%%", progressPct)
 			}
 
-			// Use Patch to avoid conflicts with main controller
-			patch := client.MergeFrom(trainJob.DeepCopy())
-			if err := UpdateTrainerStatusAnnotation(trainJob, annotationStatus); err != nil {
+			// Update annotation with termination message data
+			oldTrainJob := trainJob.DeepCopy()
+			if err := UpdateTrainerStatusAnnotation(trainJob, terminationStatus); err != nil {
 				return false, fmt.Errorf("failed to update trainer status annotation: %w", err)
 			}
+			patch := client.MergeFrom(oldTrainJob)
 			if err := c.Patch(ctx, trainJob, patch); err != nil {
 				return false, fmt.Errorf("failed to patch TrainJob annotations: %w", err)
 			}
 
 			return true, nil
+		} else {
+			log.V(1).Info("Termination message not available, trying HTTP polling", "error", termErr)
+		}
+
+		// Priority 2: Try HTTP polling (only if pod is still running with IP)
+		if pod.Status.Phase == corev1.PodRunning && pod.Status.PodIP != "" {
+			metricsPort := GetMetricsPort(trainJob)
+			if status, pollErr := PollTrainingProgress(ctx, pod, metricsPort); pollErr == nil {
+				annotationStatus := ToAnnotationStatus(status)
+				annotationStatus.LastUpdatedTime = time.Now().UTC().Format(time.RFC3339)
+
+				// Add descriptive summary
+				if completed {
+					earlyStop := false
+					if annotationStatus.CurrentStep != nil && annotationStatus.TotalSteps != nil && *annotationStatus.TotalSteps > 0 {
+						if *annotationStatus.CurrentStep < *annotationStatus.TotalSteps {
+							earlyStop = true
+						}
+					}
+					if earlyStop {
+						annotationStatus.EstimatedRemainingTimeSummary = "complete (early stopped)"
+					} else {
+						annotationStatus.EstimatedRemainingTimeSummary = "complete"
+					}
+				} else {
+					progressPct := 0
+					if annotationStatus.ProgressPercentage != nil {
+						progressPct = *annotationStatus.ProgressPercentage
+					}
+					annotationStatus.EstimatedRemainingTimeSummary = fmt.Sprintf("failed at %d%%", progressPct)
+				}
+
+				oldTrainJob := trainJob.DeepCopy()
+				if err := UpdateTrainerStatusAnnotation(trainJob, annotationStatus); err != nil {
+					return false, fmt.Errorf("failed to update trainer status annotation: %w", err)
+				}
+				patch := client.MergeFrom(oldTrainJob)
+				if err := c.Patch(ctx, trainJob, patch); err != nil {
+					return false, fmt.Errorf("failed to patch TrainJob annotations: %w", err)
+				}
```

Comment:
nit: Maybe move the status patching logic outside the if statements to avoid duplicating the logic between the termination case and the polling one.